File: PersonalDigitalVault.API/SecureVault/Services/SearchService.cs
using PersonalDigitalVault.API.Repositories.Interfaces;
using PersonalDigitalVault.API.SecureVault.DTOs;

namespace PersonalDigitalVault.API.SecureVault.Services
{
    public class SearchService : ISearchService
    {
        private readonly ISearchRepository _searchRepository;

        public SearchService(ISearchRepository searchRepository)
        {
            _searchRepository = searchRepository;
        }

        public async Task<SearchResponse> SearchAsync(
            int userId,
            string searchTerm)
        {
            if (string.IsNullOrWhiteSpace(searchTerm))
                throw new ArgumentException("Search term is required.");

            searchTerm = searchTerm.Trim();

            var folders = await _searchRepository.SearchFoldersAsync(
                userId,
                searchTerm);

            var documents = await _searchRepository.SearchDocumentsAsync(
                userId,
                searchTerm);

            var credentials = await _searchRepository.SearchCredentialsAsync(
                userId,
                searchTerm);

            return new SearchResponse
            {
                Folders = folders.Select(f => new SearchFolderResponse
                {
                    FolderId = f.FolderId,
                    FolderName = f.FolderName,
                    Description = f.Description
                }).ToList(),

                Documents = documents.Select(d => new SearchDocumentResponse
                {
                    DocumentId = d.DocumentId,
                    FolderId = d.FolderId,
                    OriginalFileName = d.OriginalFileName,
                    FileType = d.FileType
                }).ToList(),

                Credentials = credentials.Select(c =>
                    new SearchCredentialResponse
                    {
                        CredentialId = c.CredentialId,
                        FolderId = c.FolderId,
                        Title = c.Title,
                        UserName = c.UserName,
                        Notes = c.Notes
                    }).ToList()
            };
        }
    }
}

File: PersonalDigitalVault.Frontend/src/app/features/secure-vault/services/search.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../../../environments/environment';

export interface SearchFolderResult {
  folderId: number;
  folderName: string;
  description: string | null;
}

export interface SearchDocumentResult {
  documentId: number;
  folderId: number | null;
  originalFileName: string;
  fileType: string;
}

export interface SearchCredentialResult {
  credentialId: number;
  folderId: number | null;
  title: string;
  userName: string;
  notes: string | null;
}

export interface VaultSearchResults {
  folders: SearchFolderResult[];
  documents: SearchDocumentResult[];
  credentials: SearchCredentialResult[];
}

@Injectable({
  providedIn: 'root'
})
export class SearchService {
  private http = inject(HttpClient);

  private apiUrl = `${environment.apiUrl}/search`;

  // Endpoint: GET /api/search?searchTerm=query
  searchVault(searchTerm: string): Observable<VaultSearchResults> {
    return this.http.get<VaultSearchResults>(
      `${this.apiUrl}?searchTerm=${encodeURIComponent(searchTerm)}`
    );
  }
}